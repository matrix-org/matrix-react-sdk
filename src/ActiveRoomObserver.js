/*
Copyright 2017 New Vector Ltd

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import RoomViewStore from './stores/RoomViewStore';

/**
 * Consumes changes from the RoomViewStore and notifies specific things
 * about when the active room changes. Unlike listening for RoomViewStore
 * changes, you can subscribe to only changes relevant to a particular
 * room.
 *
 * TODO: If we introduce an observer for something else, factor out
 * the adding / removing of listeners & emitting into a common class.
 */
class ActiveRoomObserver {
    constructor() {
        this._listeners = {};

        this._activeRoomId = RoomViewStore.getRoomId();
        // TODO: We could self-destruct when the last listener goes away, or at least
        // stop listening.
        this._roomStoreToken = RoomViewStore.addListener(this._onRoomViewStoreUpdate.bind(this));
    }

    addListener(roomId, listener) {
        if (!this._listeners[roomId]) this._listeners[roomId] = [];
        this._listeners[roomId].push(listener);
    }

    removeListener(roomId, listener) {
        if (this._listeners[roomId]) {
            const i = this._listeners[roomId].indexOf(listener);
            if (i > -1) {
                this._listeners[roomId].splice(i, 1);
            }
        } else {
            console.warn("Unregistering unrecognised listener (roomId=" + roomId + ")");
        }
    }

    _emit(roomId) {
        if (!this._listeners[roomId]) return;

        for (const l of this._listeners[roomId]) {
            l.call();
        }
    }

    _onRoomViewStoreUpdate() {
        // emit for the old room ID
        if (this._activeRoomId) this._emit(this._activeRoomId);

        // update our cache
        this._activeRoomId = RoomViewStore.getRoomId();

        // and emit for the new one
        if (this._activeRoomId) this._emit(this._activeRoomId);
    }
}

if (global.mx_ActiveRoomObserver === undefined) {
    global.mx_ActiveRoomObserver = new ActiveRoomObserver();
}
export default global.mx_ActiveRoomObserver;
