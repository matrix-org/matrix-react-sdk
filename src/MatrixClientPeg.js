/*
Copyright 2015, 2016 OpenMarket Ltd
Copyright 2017 Vector Creations Ltd.
Copyright 2017 New Vector Ltd

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import { MatrixClient, MemoryStore } from "matrix-js-sdk";

import utils from "matrix-js-sdk/lib/utils";
import EventTimeline from "matrix-js-sdk/lib/models/event-timeline";
import EventTimelineSet from "matrix-js-sdk/lib/models/event-timeline-set";
import sdk from "./index";
import createMatrixClient from "./utils/createMatrixClient";
import SettingsStore from "./settings/SettingsStore";
import MatrixActionCreators from "./actions/MatrixActionCreators";
import Modal from "./Modal";
import { verificationMethods } from "matrix-js-sdk/lib/crypto";
import MatrixClientBackedSettingsHandler from "./settings/handlers/MatrixClientBackedSettingsHandler";
import * as StorageManager from "./utils/StorageManager";
import IdentityAuthClient from "./IdentityAuthClient";

interface MatrixClientCreds {
    homeserverUrl: string;
    identityServerUrl: string;
    userId: string;
    deviceId: string;
    accessToken: string;
    guest: boolean;
}

/**
 * Wrapper object for handling the js-sdk Matrix Client object in the react-sdk
 * Handles the creation/initialisation of client objects.
 * This module provides a singleton instance of this class so the 'current'
 * Matrix Client object is available easily.
 */
class MatrixClientPeg {
    constructor() {
        this.matrixClient = null;
        this._justRegisteredUserId = null;

        // These are the default options used when when the
        // client is started in 'start'. These can be altered
        // at any time up to after the 'will_start_client'
        // event is finished processing.
        this.opts = {
            initialSyncLimit: 20
        };
        // the credentials used to init the current client object.
        // used if we tear it down & recreate it with a different store
        this._currentClientCreds = null;
    }

    /**
     * Sets the script href passed to the IndexedDB web worker
     * If set, a separate web worker will be started to run the IndexedDB
     * queries on.
     *
     * @param {string} script href to the script to be passed to the web worker
     */
    setIndexedDbWorkerScript(script) {
        createMatrixClient.indexedDbWorkerScript = script;
    }

    get(): MatrixClient {
        return this.matrixClient;
    }

    unset() {
        this.matrixClient = null;

        MatrixActionCreators.stop();
    }

    /**
     * If we've registered a user ID we set this to the ID of the
     * user we've just registered. If they then go & log in, we
     * can send them to the welcome user (obviously this doesn't
     * guarentee they'll get a chat with the welcome user).
     *
     * @param {string} uid The user ID of the user we've just registered
     */
    setJustRegisteredUserId(uid) {
        this._justRegisteredUserId = uid;
    }

    /**
     * Returns true if the current user has just been registered by this
     * client as determined by setJustRegisteredUserId()
     *
     * @returns {bool} True if user has just been registered
     */
    currentUserIsJustRegistered() {
        return (
            this.matrixClient &&
            this.matrixClient.credentials.userId === this._justRegisteredUserId
        );
    }

    /*
     * Replace this MatrixClientPeg's client with a client instance that has
     * homeserver / identity server URLs and active credentials
     */
    replaceUsingCreds(creds: MatrixClientCreds) {
        this._currentClientCreds = creds;
        this._createClient(creds);
    }

    async assign() {
        for (const dbType of ["indexeddb", "memory"]) {
            try {
                const promise = this.matrixClient.store.startup();
                console.log(
                    "MatrixClientPeg: waiting for MatrixClient store to initialise"
                );
                await promise;
                break;
            } catch (err) {
                if (dbType === "indexeddb") {
                    console.error(
                        "Error starting matrixclient store - falling back to memory store",
                        err
                    );
                    this.matrixClient.store = new MemoryStore({
                        localStorage: global.localStorage
                    });
                } else {
                    console.error("Failed to start memory store!", err);
                    throw err;
                }
            }
        }

        StorageManager.trackStores(this.matrixClient);

        // try to initialise e2e on the new client
        try {
            // check that we have a version of the js-sdk which includes initCrypto
            if (
                !SettingsStore.getValue("lowBandwidth") &&
                this.matrixClient.initCrypto
            ) {
                await this.matrixClient.initCrypto();
                StorageManager.setCryptoInitialised(true);
            }
        } catch (e) {
            if (e && e.name === "InvalidCryptoStoreError") {
                // The js-sdk found a crypto DB too new for it to use
                const CryptoStoreTooNewDialog = sdk.getComponent(
                    "views.dialogs.CryptoStoreTooNewDialog"
                );
                Modal.createDialog(CryptoStoreTooNewDialog, {
                    host: window.location.host
                });
            }
            // this can happen for a number of reasons, the most likely being
            // that the olm library was missing. It's not fatal.
            console.warn("Unable to initialise e2e", e);
        }

        const opts = utils.deepCopy(this.opts);
        // the react sdk doesn't work without this, so don't allow
        opts.pendingEventOrdering = "detached";
        opts.lazyLoadMembers = true;

        // Connect the matrix client to the dispatcher and setting handlers
        MatrixActionCreators.start(this.matrixClient);
        MatrixClientBackedSettingsHandler.matrixClient = this.matrixClient;

        return opts;
    }

    async start() {
        const opts = await this.assign();

        console.log(`MatrixClientPeg: really starting MatrixClient`);
        await this.get().startClient(opts);
        console.log(`MatrixClientPeg: MatrixClient started`);
    }

    getCredentials(): MatrixClientCreds {
        return {
            homeserverUrl: this.matrixClient.baseUrl,
            identityServerUrl: this.matrixClient.idBaseUrl,
            userId: this.matrixClient.credentials.userId,
            deviceId: this.matrixClient.getDeviceId(),
            accessToken: this.matrixClient.getAccessToken(),
            guest: this.matrixClient.isGuest()
        };
    }

    /*
     * Return the server name of the user's homeserver
     * Throws an error if unable to deduce the homeserver name
     * (eg. if the user is not logged in)
     */
    getHomeserverName() {
        const matches = /^@.+:(.+)$/.exec(this.matrixClient.credentials.userId);
        if (matches === null || matches.length < 1) {
            throw new Error("Failed to derive homeserver name from user ID!");
        }
        return matches[1];
    }

    _createClient(creds: MatrixClientCreds) {
        const opts = {
            baseUrl: creds.homeserverUrl,
            idBaseUrl: creds.identityServerUrl,
            accessToken: creds.accessToken,
            userId: creds.userId,
            deviceId: creds.deviceId,
            timelineSupport: true,
            forceTURN: !SettingsStore.getValue("webRtcAllowPeerToPeer", false),
            fallbackICEServerAllowed: !!SettingsStore.getValue(
                "fallbackICEServerAllowed"
            ),
            verificationMethods: [verificationMethods.SAS],
            unstableClientRelationAggregation: true,
            identityServer: new IdentityAuthClient()
        };

        this.matrixClient = createMatrixClient(opts);

        // we're going to add eventlisteners for each matrix event tile, so the
        // potential number of event listeners is quite high.
        this.matrixClient.setMaxListeners(500);

        this.matrixClient.setGuest(Boolean(creds.guest));

        const notifTimelineSet = new EventTimelineSet(null, {
            timelineSupport: true
        });
        // XXX: what is our initial pagination token?! it somehow needs to be synchronised with /sync.
        notifTimelineSet
            .getLiveTimeline()
            .setPaginationToken("", EventTimeline.BACKWARDS);
        this.matrixClient.setNotifTimelineSet(notifTimelineSet);
        //console.log("OPTIONS TO GET CLIENT CREDENTIALS", opts);
    }
}

if (!global.mxMatrixClientPeg) {
    global.mxMatrixClientPeg = new MatrixClientPeg();
}
export default global.mxMatrixClientPeg;
