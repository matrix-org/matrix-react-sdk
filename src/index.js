/*
Copyright 2015, 2016 OpenMarket Ltd

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

var Skinner = require('./Skinner');

var t = require('counterpart');

t.registerTranslations('en', require('../languages/en'));
t.registerTranslations('pt-BR', require('../languages/pt-BR'));

t.setFallbackLocale('en');
t.setLocale(window.navigator.languages[0]);



module.exports.loadSkin = function(skinObject) {
    Skinner.load(skinObject);
};

module.exports.resetSkin = function() {
    Skinner.reset();
};

module.exports.getComponent = function(componentName) {
    return Skinner.getComponent(componentName);
};
