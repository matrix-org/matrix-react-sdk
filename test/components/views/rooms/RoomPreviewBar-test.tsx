/*
Copyright 2021 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import React from 'react';
import {
    renderIntoDocument,
    Simulate,
    findRenderedDOMComponentWithClass,
    act,
} from 'react-dom/test-utils';
import { Room, RoomMember, MatrixError, IContent } from 'matrix-js-sdk/src/matrix';

import { stubClient } from '../../../test-utils';
import { MatrixClientPeg } from '../../../../src/MatrixClientPeg';
import DMRoomMap from '../../../../src/utils/DMRoomMap';
import RoomPreviewBar from '../../../../src/components/views/rooms/RoomPreviewBar';

jest.mock('../../../../src/IdentityAuthClient', () => {
    return jest.fn().mockImplementation(() => {
        return { getAccessToken: jest.fn().mockResolvedValue('mock-token') };
    });
});

jest.useRealTimers();

const createRoom = (roomId: string, userId: string): Room => {
    const newRoom = new Room(
        roomId,
        MatrixClientPeg.get(),
        userId,
        {},
    );
    DMRoomMap.makeShared().start();
    return newRoom;
};

const makeMockRoomMember = (
    { userId, isKicked, membership, content, memberContent }:
    {userId?: string;
        isKicked?: boolean;
        membership?: 'invite' | 'ban';
        content?: Partial<IContent>;
        memberContent?: Partial<IContent>;
    },
) => ({
    userId,
    rawDisplayName: `${userId} name`,
    isKicked: jest.fn().mockReturnValue(!!isKicked),
    getContent: jest.fn().mockReturnValue(content || {}),
    membership,
    events: {
        member: {
            getSender: jest.fn().mockReturnValue('@kicker:test.com'),
            getContent: jest.fn().mockReturnValue({ reason: 'test reason', ...memberContent }),
        },
    },
}) as unknown as RoomMember;

describe('<RoomPreviewBar />', () => {
    const roomId = 'RoomPreviewBar-test-room';
    const userId = '@tester:test.com';
    const inviterUserId = '@inviter:test.com';
    const otherUserId = '@othertester:test.com';

    const getComponent = (props = {}) => {
        const defaultProps = {
            room: createRoom(roomId, userId),
        };
        const wrapper = renderIntoDocument<React.Component>(
            <RoomPreviewBar {...defaultProps} {...props} />,
        ) as React.Component;
        return findRenderedDOMComponentWithClass(wrapper, 'mx_RoomPreviewBar') as HTMLDivElement;
    };

    const isSpinnerRendered = (element: Element) => !!element.querySelector('.mx_Spinner');
    const getMessage = (element: Element) => element.querySelector<HTMLDivElement>('.mx_RoomPreviewBar_message');
    const getActions = (element: Element) => element.querySelector<HTMLDivElement>('.mx_RoomPreviewBar_actions');
    const getPrimaryActionButton = (element: Element) =>
        getActions(element).querySelector('.mx_AccessibleButton_kind_primary');
    const getSecondaryActionButton = (element: Element) =>
        getActions(element).querySelector('.mx_AccessibleButton_kind_secondary');

    beforeEach(() => {
        stubClient();
        MatrixClientPeg.get().getUserId = jest.fn().mockReturnValue(userId);
    });

    afterEach(() => {
        const container = document.body.firstChild;
        container && document.body.removeChild(container);
    });

    it('renders joining message', () => {
        const component = getComponent({ joining: true });

        expect(isSpinnerRendered(component)).toBeTruthy();
        expect(getMessage(component).textContent).toEqual('Joining …');
    });
    it('renders rejecting message', () => {
        const component = getComponent({ rejecting: true });
        expect(isSpinnerRendered(component)).toBeTruthy();
        expect(getMessage(component).textContent).toEqual('Rejecting invite …');
    });
    it('renders loading message', () => {
        const component = getComponent({ loading: true });
        expect(isSpinnerRendered(component)).toBeTruthy();
        expect(getMessage(component).textContent).toEqual('Loading …');
    });

    it('renders not logged in message', () => {
        MatrixClientPeg.get().isGuest = jest.fn().mockReturnValue(true);
        const component = getComponent({ loading: true });

        expect(isSpinnerRendered(component)).toBeFalsy();
        expect(getMessage(component).textContent).toEqual('Join the conversation with an account');
    });

    it('renders kicked message', () => {
        const room = createRoom(roomId, otherUserId);
        jest.spyOn(room, 'getMember').mockReturnValue(makeMockRoomMember({ isKicked: true }));
        const component = getComponent({ loading: true, room });

        expect(getMessage(component)).toMatchSnapshot();
    });

    it('renders banned message', () => {
        const room = createRoom(roomId, otherUserId);
        jest.spyOn(room, 'getMember').mockReturnValue(makeMockRoomMember({ membership: 'ban' }));
        const component = getComponent({ loading: true, room });

        expect(getMessage(component)).toMatchSnapshot();
    });

    describe('with an error', () => {
        it('renders room not found error', () => {
            const error = new MatrixError({
                errcode: 'M_NOT_FOUND',
                error: "Room not found",
            });
            const component = getComponent({ error });

            expect(getMessage(component)).toMatchSnapshot();
        });
        it('renders other errors', () => {
            const error = new MatrixError({
                errcode: 'Something_else',
            });
            const component = getComponent({ error });

            expect(getMessage(component)).toMatchSnapshot();
        });
    });

    it('renders viewing room message when room an be previewed', () => {
        const component = getComponent({ canPreview: true });

        expect(getMessage(component)).toMatchSnapshot();
    });

    it('renders viewing room message when room can not be previewed', () => {
        const component = getComponent({ canPreview: false });

        expect(getMessage(component)).toMatchSnapshot();
    });

    describe('with an invite', () => {
        const inviterName = inviterUserId;
        const userMember = makeMockRoomMember({ userId });
        const userMemberWithDmInvite = makeMockRoomMember({
            userId, membership: 'invite', memberContent: { is_direct: true },
        });
        const inviterMember = makeMockRoomMember({
            userId: inviterUserId,
            content: {
                "reason": 'test',
                'io.element.html_reason': '<h3>hello</h3>',
            },
        });
        describe('without an invited email', () => {
            describe('for a non-dm room', () => {
                const mockGetMember = (id) => {
                    if (id === userId) return userMember;
                    return inviterMember;
                };
                const onJoinClick = jest.fn();
                const onRejectClick = jest.fn();
                let room;

                beforeEach(() => {
                    room = createRoom(roomId, userId);
                    jest.spyOn(room, 'getMember').mockImplementation(mockGetMember);
                    jest.spyOn(room.currentState, 'getMember').mockImplementation(mockGetMember);
                    onJoinClick.mockClear();
                    onRejectClick.mockClear();
                });

                it('renders invite message', () => {
                    const component = getComponent({ inviterName, room });
                    expect(getMessage(component)).toMatchSnapshot();
                });

                it('renders join and reject action buttons correctly', () => {
                    const component = getComponent({ inviterName, room, onJoinClick, onRejectClick });
                    expect(getActions(component)).toMatchSnapshot();
                });

                it('renders reject and ignore action buttons when handler is provided', () => {
                    const onRejectAndIgnoreClick = jest.fn();
                    const component = getComponent({
                        inviterName, room, onJoinClick, onRejectClick, onRejectAndIgnoreClick,
                    });
                    expect(getActions(component)).toMatchSnapshot();
                });

                it('renders join and reject action buttons in reverse order when room can previewed', () => {
                    // when room is previewed action buttons are rendered left to right, with primary on the right
                    const component = getComponent({ inviterName, room, onJoinClick, onRejectClick, canPreview: true });
                    expect(getActions(component)).toMatchSnapshot();
                });

                it('joins room on primary button click', () => {
                    const component = getComponent({ inviterName, room, onJoinClick, onRejectClick });
                    act(() => {
                        Simulate.click(getPrimaryActionButton(component));
                    });

                    expect(onJoinClick).toHaveBeenCalled();
                });

                it('rejects invite on secondary button click', () => {
                    const component = getComponent({ inviterName, room, onJoinClick, onRejectClick });
                    act(() => {
                        Simulate.click(getSecondaryActionButton(component));
                    });

                    expect(onRejectClick).toHaveBeenCalled();
                });
            });

            describe('for a dm room', () => {
                const mockGetMember = (id) => {
                    if (id === userId) return userMemberWithDmInvite;
                    return inviterMember;
                };
                const onJoinClick = jest.fn();
                const onRejectClick = jest.fn();
                let room;

                beforeEach(() => {
                    room = createRoom(roomId, userId);
                    jest.spyOn(room, 'getMember').mockImplementation(mockGetMember);
                    jest.spyOn(room.currentState, 'getMember').mockImplementation(mockGetMember);
                    onJoinClick.mockClear();
                    onRejectClick.mockClear();
                });

                it('renders invite message to a non-dm room', () => {
                    const component = getComponent({ inviterName, room });
                    expect(getMessage(component)).toMatchSnapshot();
                });

                it('renders join and reject action buttons with correct labels', () => {
                    const onRejectAndIgnoreClick = jest.fn();
                    const component = getComponent({
                        inviterName, room, onJoinClick, onRejectAndIgnoreClick, onRejectClick,
                    });
                    expect(getActions(component)).toMatchSnapshot();
                });
            });
        });

        describe('with an invited email', () => {
            const invitedEmail = 'test@test.com';
            const mockThreePids = [
                { medium: 'email', address: invitedEmail },
                { medium: 'not-email', address: 'address 2' },
            ];

            const testJoinButton = (props) => async () => {
                const onJoinClick = jest.fn();
                const onRejectClick = jest.fn();
                const component = getComponent({ ...props, onJoinClick, onRejectClick });
                await new Promise(setImmediate);
                expect(getPrimaryActionButton(component)).toBeTruthy();
                expect(getSecondaryActionButton(component)).toBeFalsy();
                act(() => {
                    Simulate.click(getPrimaryActionButton(component));
                });
                expect(onJoinClick).toHaveBeenCalled();
            };

            describe('when client fails to get 3PIDs', () => {
                beforeEach(() => {
                    MatrixClientPeg.get().getThreePids = jest.fn().mockRejectedValue({ errCode: 'TEST_ERROR' });
                });

                it('renders error message', async () => {
                    const component = getComponent({ inviterName, invitedEmail });
                    await new Promise(setImmediate);

                    expect(getMessage(component)).toMatchSnapshot();
                });

                it('renders join button', testJoinButton({ inviterName, invitedEmail }));
            });

            describe('when invitedEmail is not associated with current account', () => {
                beforeEach(() => {
                    MatrixClientPeg.get().getThreePids = jest.fn().mockResolvedValue(
                        { threepids: mockThreePids.slice(1) },
                    );
                });

                it('renders invite message with invited email', async () => {
                    const component = getComponent({ inviterName, invitedEmail });
                    await new Promise(setImmediate);

                    expect(getMessage(component)).toMatchSnapshot();
                });

                it('renders join button', testJoinButton({ inviterName, invitedEmail }));
            });

            describe('when client has no identity server connected', () => {
                beforeEach(() => {
                    MatrixClientPeg.get().getThreePids = jest.fn().mockResolvedValue({ threepids: mockThreePids });
                    MatrixClientPeg.get().getIdentityServerUrl = jest.fn().mockReturnValue(false);
                });

                it('renders invite message with invited email', async () => {
                    const component = getComponent({ inviterName, invitedEmail });
                    await new Promise(setImmediate);

                    expect(getMessage(component)).toMatchSnapshot();
                });

                it('renders join button', testJoinButton({ inviterName, invitedEmail }));
            });

            describe('when client has an identity server connected', () => {
                beforeEach(() => {
                    MatrixClientPeg.get().getThreePids = jest.fn().mockResolvedValue({ threepids: mockThreePids });
                    MatrixClientPeg.get().getIdentityServerUrl = jest.fn().mockReturnValue('identity.test');
                    MatrixClientPeg.get().lookupThreePid = jest.fn().mockResolvedValue('identity.test');
                });

                it('renders email mismatch message when invite email mxid doesnt match', async () => {
                    MatrixClientPeg.get().lookupThreePid = jest.fn().mockReturnValue('not userid');
                    const component = getComponent({ inviterName, invitedEmail });
                    await new Promise(setImmediate);

                    expect(getMessage(component)).toMatchSnapshot();
                    expect(MatrixClientPeg.get().lookupThreePid).toHaveBeenCalledWith(
                        'email', invitedEmail, undefined, 'mock-token',
                    );
                    await testJoinButton({ inviterName, invitedEmail })();
                });

                it('renders invite message when invite email mxid match', async () => {
                    MatrixClientPeg.get().lookupThreePid = jest.fn().mockReturnValue(userId);
                    const component = getComponent({ inviterName, invitedEmail });
                    await new Promise(setImmediate);

                    expect(getMessage(component)).toMatchSnapshot();
                    await testJoinButton({ inviterName, invitedEmail })();
                });
            });
        });
    });
});
