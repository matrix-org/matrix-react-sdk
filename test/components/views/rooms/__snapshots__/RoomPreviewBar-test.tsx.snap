// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RoomPreviewBar /> renders banned message 1`] = `
<div
  class="mx_RoomPreviewBar_message"
>
  <h3>
    You were banned from RoomPreviewBar-test-room by @kicker:test.com
  </h3>
  <p>
    Reason: test reason
  </p>
</div>
`;

exports[`<RoomPreviewBar /> renders kicked message 1`] = `
<div
  class="mx_RoomPreviewBar_message"
>
  <h3>
    You were removed from RoomPreviewBar-test-room by @kicker:test.com
  </h3>
  <p>
    Reason: test reason
  </p>
</div>
`;

exports[`<RoomPreviewBar /> renders viewing room message when room an be previewed 1`] = `
<div
  class="mx_RoomPreviewBar_message"
>
  <h3>
    You're previewing RoomPreviewBar-test-room. Want to join it?
  </h3>
</div>
`;

exports[`<RoomPreviewBar /> renders viewing room message when room can not be previewed 1`] = `
<div
  class="mx_RoomPreviewBar_message"
>
  <h3>
    RoomPreviewBar-test-room can't be previewed. Do you want to join it?
  </h3>
</div>
`;

exports[`<RoomPreviewBar /> with an error renders other errors 1`] = `
<div
  class="mx_RoomPreviewBar_message"
>
  <h3>
    RoomPreviewBar-test-room is not accessible at this time.
  </h3>
  <p>
    Try again later, or ask a room or space admin to check if you have access.
  </p>
  <p>
    <span>
      Something_else was returned while trying to access the room or space. If you think you're seeing this message in error, please 
      <a
        href="https://github.com/vector-im/element-web/issues/new/choose"
        rel="noreferrer noopener"
        target="_blank"
      >
        submit a bug report
      </a>
      .
    </span>
  </p>
</div>
`;

exports[`<RoomPreviewBar /> with an error renders room not found error 1`] = `
<div
  class="mx_RoomPreviewBar_message"
>
  <h3>
    RoomPreviewBar-test-room does not exist.
  </h3>
  <p>
    Are you sure you're at the right place?
  </p>
</div>
`;

exports[`<RoomPreviewBar /> with an invite with an invited email when client fails to get 3PIDs renders error message 1`] = `
<div
  class="mx_RoomPreviewBar_message"
>
  <h3>
    Something went wrong with your invite to RoomPreviewBar-test-room
  </h3>
  <p>
    An error (unknown error code) was returned while trying to validate your invite. You could try to pass this information on to the person who invited you.
  </p>
</div>
`;

exports[`<RoomPreviewBar /> with an invite with an invited email when client has an identity server connected renders email mismatch message when invite email mxid doesnt match 1`] = `
<div
  class="mx_RoomPreviewBar_message"
>
  <h3>
    This invite to RoomPreviewBar-test-room was sent to test@test.com
  </h3>
  <p>
    Share this email in Settings to receive invites directly in Element.
  </p>
</div>
`;

exports[`<RoomPreviewBar /> with an invite with an invited email when client has an identity server connected renders invite message when invite email mxid match 1`] = `
<div
  class="mx_RoomPreviewBar_message"
>
  <h3>
    This invite to RoomPreviewBar-test-room was sent to test@test.com
  </h3>
  <p>
    Share this email in Settings to receive invites directly in Element.
  </p>
</div>
`;

exports[`<RoomPreviewBar /> with an invite with an invited email when client has no identity server connected renders invite message with invited email 1`] = `
<div
  class="mx_RoomPreviewBar_message"
>
  <h3>
    This invite to RoomPreviewBar-test-room was sent to test@test.com
  </h3>
  <p>
    Use an identity server in Settings to receive invites directly in Element.
  </p>
</div>
`;

exports[`<RoomPreviewBar /> with an invite with an invited email when invitedEmail is not associated with current account renders invite message with invited email 1`] = `
<div
  class="mx_RoomPreviewBar_message"
>
  <h3>
    This invite to RoomPreviewBar-test-room was sent to test@test.com which is not associated with your account
  </h3>
  <p>
    Link this email with your account in Settings to receive invites directly in Element.
  </p>
</div>
`;

exports[`<RoomPreviewBar /> with an invite without an invited email for a dm room renders invite message to a non-dm room 1`] = `
<div
  class="mx_RoomPreviewBar_message"
>
  <h3>
    Do you want to join RoomPreviewBar-test-room?
  </h3>
  <p>
    <span
      class="mx_BaseAvatar"
      role="presentation"
    >
      <span
        aria-hidden="true"
        class="mx_BaseAvatar_initial"
        style="font-size: 23.400000000000002px; width: 36px; line-height: 36px;"
      >
        R
      </span>
      <img
        alt=""
        aria-hidden="true"
        class="mx_BaseAvatar_image"
        src="data:image/png;base64,00"
        style="width: 36px; height: 36px;"
      />
    </span>
  </p>
  <p>
    <span>
      <span>
        <span
          class="mx_RoomPreviewBar_inviter"
        >
          @inviter:test.com name
        </span>
         (
        @inviter:test.com
        )
      </span>
       invited you
    </span>
  </p>
</div>
`;

exports[`<RoomPreviewBar /> with an invite without an invited email for a dm room renders join and reject action buttons with correct labels 1`] = `
<div
  class="mx_RoomPreviewBar_actions"
>
  <div
    class="mx_AccessibleButton mx_AccessibleButton_hasKind mx_AccessibleButton_kind_primary"
    role="button"
    tabindex="0"
  >
    Accept
  </div>
  <div
    class="mx_AccessibleButton mx_AccessibleButton_hasKind mx_AccessibleButton_kind_secondary"
    role="button"
    tabindex="0"
  >
    Reject & Ignore user
  </div>
  <div
    class="mx_AccessibleButton mx_AccessibleButton_hasKind mx_AccessibleButton_kind_secondary"
    role="button"
    tabindex="0"
  >
    Reject
  </div>
</div>
`;

exports[`<RoomPreviewBar /> with an invite without an invited email for a non-dm room renders invite message 1`] = `
<div
  class="mx_RoomPreviewBar_message"
>
  <h3>
    Do you want to join RoomPreviewBar-test-room?
  </h3>
  <p>
    <span
      class="mx_BaseAvatar"
      role="presentation"
    >
      <span
        aria-hidden="true"
        class="mx_BaseAvatar_initial"
        style="font-size: 23.400000000000002px; width: 36px; line-height: 36px;"
      >
        R
      </span>
      <img
        alt=""
        aria-hidden="true"
        class="mx_BaseAvatar_image"
        src="data:image/png;base64,00"
        style="width: 36px; height: 36px;"
      />
    </span>
  </p>
  <p>
    <span>
      <span>
        <span
          class="mx_RoomPreviewBar_inviter"
        >
          @inviter:test.com name
        </span>
         (
        @inviter:test.com
        )
      </span>
       invited you
    </span>
  </p>
</div>
`;

exports[`<RoomPreviewBar /> with an invite without an invited email for a non-dm room renders join and reject action buttons correctly 1`] = `
<div
  class="mx_RoomPreviewBar_actions"
>
  <div
    class="mx_AccessibleButton mx_AccessibleButton_hasKind mx_AccessibleButton_kind_primary"
    role="button"
    tabindex="0"
  >
    Accept
  </div>
  <div
    class="mx_AccessibleButton mx_AccessibleButton_hasKind mx_AccessibleButton_kind_secondary"
    role="button"
    tabindex="0"
  >
    Reject
  </div>
</div>
`;

exports[`<RoomPreviewBar /> with an invite without an invited email for a non-dm room renders join and reject action buttons in reverse order when room can previewed 1`] = `
<div
  class="mx_RoomPreviewBar_actions"
>
  <div
    class="mx_AccessibleButton mx_AccessibleButton_hasKind mx_AccessibleButton_kind_secondary"
    role="button"
    tabindex="0"
  >
    Reject
  </div>
  <div
    class="mx_AccessibleButton mx_AccessibleButton_hasKind mx_AccessibleButton_kind_primary"
    role="button"
    tabindex="0"
  >
    Accept
  </div>
</div>
`;

exports[`<RoomPreviewBar /> with an invite without an invited email for a non-dm room renders reject and ignore action buttons when handler is provided 1`] = `
<div
  class="mx_RoomPreviewBar_actions"
>
  <div
    class="mx_AccessibleButton mx_AccessibleButton_hasKind mx_AccessibleButton_kind_primary"
    role="button"
    tabindex="0"
  >
    Accept
  </div>
  <div
    class="mx_AccessibleButton mx_AccessibleButton_hasKind mx_AccessibleButton_kind_secondary"
    role="button"
    tabindex="0"
  >
    Reject & Ignore user
  </div>
  <div
    class="mx_AccessibleButton mx_AccessibleButton_hasKind mx_AccessibleButton_kind_secondary"
    role="button"
    tabindex="0"
  >
    Reject
  </div>
</div>
`;
